File: packages/server/src/utils/get-full-project-summary.ts
import { ApiError } from '@octoprompt/shared'
import { ProjectFile } from '@octoprompt/schemas'
import { getProjectById, getProjectFiles } from '@/services/project-service'
import { buildCombinedFileSummariesXml } from 'shared/src/utils/summary-formatter'

const buildProjectSummary = (includedFiles: ProjectFile[]) => {
  // Build the combined summaries using your summary-formatter
  return buildCombinedFileSummariesXml(includedFiles, {
    includeEmptySummaries: true
  })
}

export const getFullProjectSummary = async (projectId: number) => {
  const project = await getProjectById(projectId)
  if (!project) {
    throw new ApiError(404, 'Project not found', 'NOT_FOUND')
  }

  // Fetch all file summaries from the database
  const allFiles = (await getProjectFiles(projectId)) || []
  if (!allFiles.length) {
    return {
      success: false,
      message: 'No summaries available. Please summarize files first.'
    }
  }

  // Retrieve global state to get ignore patterns (or other filtering preferences)
  // const ignorePatterns = globalState.settings.summarizationIgnorePatterns || [];
  // const allowPatterns = globalState.settings.summarizationAllowPatterns || [];

  // Filter out files that match ignore patterns (unless a file also matches an allow pattern, if applicable)
  // The same logic your summarization page uses can be applied here:
  function isIncluded(file: ProjectFile): boolean {
    // If any ignore pattern matches, we skip—unless an allow pattern overrides it.
    // const matchesIgnore = matchesAnyPattern(file.path, ignorePatterns);
    // if (matchesIgnore && !matchesAnyPattern(file.path, allowPatterns)) {
    //     return false;
    // }
    return true
  }

  // Filter down to the "included" files
  const includedFiles = allFiles.filter(isIncluded)

  return buildProjectSummary(includedFiles)
}

File: packages/python_backend/app/utils/get_full_project_summary.py
from typing import List, Dict, Any, Union
from app.error_handling.api_error import ApiError
from app.schemas.project_schemas import Project, ProjectFile

def build_combined_file_summaries_xml(files: List[ProjectFile], options: Dict[str, Any]) -> str:
    if not files:
        return "<summaries></summaries>"
    
    include_empty = options.get("includeEmptySummaries", True)
    xml_parts = ["<summaries>"]
    
    for file_item in files:
        summary_content = file_item.summary if file_item.summary else ""
        
        if not include_empty and not summary_content:
            continue
            
        xml_parts.append("  <file>")
        xml_parts.append(f"    <path>{file_item.path}</path>")
        if include_empty or summary_content:
            xml_parts.append(f"    <summary>{summary_content}</summary>")
        xml_parts.append("  </file>")
    
    xml_parts.append("</summaries>")
    return "\n".join(xml_parts)

def build_project_summary(included_files: List[ProjectFile]) -> str:
    return build_combined_file_summaries_xml(included_files, {
        "includeEmptySummaries": True
    })

async def get_full_project_summary(project_id: int) -> Union[str, Dict[str, Any]]:
    from app.services.project_service import get_project_by_id, get_project_files
    
    project = await get_project_by_id(project_id)
    if not project:
        raise ApiError(404, "Project not found", "NOT_FOUND")

    all_files = await get_project_files(project_id)
    if not all_files:
        return {
            "success": False,
            "message": "No summaries available. Please summarize files first."
        }

    def is_included(file: ProjectFile) -> bool:
        return True

    included_files = [f for f in all_files if is_included(f)]

    return build_project_summary(included_files)